File: mobile/concord/lib/api/friends_service.dart
import 'api_client.dart';
import 'models/friend_request_summary.dart';
import 'models/paged_result.dart';
import 'models/public_profile_response.dart';
import 'models/send_request_response.dart';

class FriendsService {
  FriendsService(this._client);

  final ApiClient _client;

  Future<SendRequestResponse> sendRequest(String targetUserId) async {
    final data = await _client.post('/Friends/Requests/$targetUserId');
    return SendRequestResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<void> acceptRequest(String requestId) => _client.post('/Friends/Requests/$requestId:Accept');

  Future<void> declineRequest(String requestId) => _client.post('/Friends/Requests/$requestId:Decline');

  Future<void> cancelRequest(String requestId) => _client.delete('/Friends/Requests/$requestId');

  Future<List<FriendRequestSummary>> incomingRequests() async {
    final data = await _client.get('/Friends/Requests/Incoming');
    return (data as List<dynamic>)
        .map((e) => FriendRequestSummary.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<FriendRequestSummary>> outgoingRequests() async {
    final data = await _client.get('/Friends/Requests/Outgoing');
    return (data as List<dynamic>)
        .map((e) => FriendRequestSummary.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<PagedResult<PublicProfileResponse>> listFriends({int page = 1, int pageSize = 30}) async {
    final data = await _client.get('/Friends', query: {'page': page, 'pageSize': pageSize});
    return PagedResult.fromJson(data as Map<String, dynamic>, PublicProfileResponse.fromJson);
  }

  Future<void> block(String targetUserId) => _client.post('/Friends/Blocks/$targetUserId');

  Future<void> unblock(String targetUserId) => _client.delete('/Friends/Blocks/$targetUserId');

  Future<PagedResult<PublicProfileResponse>> listBlocked({int page = 1, int pageSize = 30}) async {
    final data = await _client.get('/Friends/Blocks', query: {'page': page, 'pageSize': pageSize});
    return PagedResult.fromJson(data as Map<String, dynamic>, PublicProfileResponse.fromJson);
  }
}

File: frontend/Concord/src/api/friendsService.js
import { request } from './httpClient';
import { buildPagingQuery } from './pagination';

export async function sendFriendRequest(targetUserId) {
  return request(`Friends/Requests/${targetUserId}`, { method: 'POST' });
}

export async function acceptFriendRequest(requestId) {
  await request(`Friends/Requests/${requestId}:Accept`, { method: 'POST' });
}

export async function declineFriendRequest(requestId) {
  await request(`Friends/Requests/${requestId}:Decline`, { method: 'POST' });
}

export async function cancelFriendRequest(requestId) {
  await request(`Friends/Requests/${requestId}`, { method: 'DELETE' });
}

export async function getIncomingFriendRequests() {
  return request('Friends/Requests/Incoming');
}

export async function getOutgoingFriendRequests() {
  return request('Friends/Requests/Outgoing');
}

export async function getFriends({ page, pageSize } = {}) {
  return request('Friends', { query: buildPagingQuery(page, pageSize) });
}

export async function blockUser(targetUserId) {
  await request(`Friends/Blocks/${targetUserId}`, { method: 'POST' });
}

export async function unblockUser(targetUserId) {
  await request(`Friends/Blocks/${targetUserId}`, { method: 'DELETE' });
}

export async function getBlockedUsers({ page, pageSize } = {}) {
  return request('Friends/Blocks', { query: buildPagingQuery(page, pageSize) });
}
